pysrc/bytewax/connectors/kafka/registry.py

```diff
@@ -0,0 +1,222 @@
+"""TODO."""
+import logging
+from abc import ABC, abstractmethod
+from dataclasses import dataclass
+from typing import Dict, Optional
+
+import requests
+from confluent_kafka.schema_registry import SchemaRegistryClient
+
+from .serde import (
+    SchemaDeserializer,
+    SchemaSerializer,
+    _AvroDeserializer,
+    _AvroSerializer,
+    _ConfluentAvroDeserializer,
+    _ConfluentAvroSerializer,
+)
+
+logger = logging.getLogger(__name__)
+
+
+@dataclass(frozen=True)
+class SchemaConf:
+    """Info used to retrieve a schema from a schema registry.
+
+    Either use `schema_id`, or specify the `subject` and optionally `version`.
+    If no `version` is specified, it defaults to the latest schema.
+    """
+
+    # TODO: Only `avro` supported for now, but we might want
+    #       to add `protobuf` and others too
+    # format: str = "avro"
+    schema_id: Optional[int] = None
+    subject: Optional[str] = None
+    version: Optional[int] = None
+
+    def __post_init__(self):
+        """Validate the data."""
+        if self.schema_id is None:
+            if self.subject is None:
+                msg = "subject MUST be specified if no schema_id is provided"
+                raise ValueError(msg)
+        else:
+            if self.subject is not None:
+                logger.warning(
+                    "both schema_id and subject specified, subject will be ignored"
+                )
+
+
+class SchemaRegistry(ABC):
+    """An interface to define a SchemaRegistry.
+
+    A schema registry must define functions to build
+    a (de)serializer for both key and value of a KafkaMessage.
+    """
+
+    @abstractmethod
+    def key_serializer(self) -> SchemaSerializer:
+        """Return a serializer for keys."""
+        ...
+
+    @abstractmethod
+    def value_serializer(self) -> SchemaSerializer:
+        """Return a serializer for values."""
+        ...
+
+    @abstractmethod
+    def key_deserializer(self) -> SchemaDeserializer:
+        """Return a deserializer for keys."""
+        ...
+
+    @abstractmethod
+    def value_deserializer(self) -> SchemaDeserializer:
+        """Return a deserializer for values."""
+        ...
+
+
+class ConfluentSchemaRegistry(SchemaRegistry):
+    """Confluent's schema registry for Kafka's input and output connectors.
+
+    Serializer: Use either `schema_id` or `subject`+`version` to fetch
+    the schema for the serializer from the registry.
+
+    Deserializer: The deserializer automatically fetches the correct
+    schema for each message, following confluent_kafka's behaviour
+    """
+
+    def __init__(
+        self,
+        sr_conf: Dict,
```

Comment:
The problem here is that both confluent's serializer and deserializer require a full instance of a client when calling the (de)serialization function. So I only pass the conf to instantiate a client for each (de)serializer instance. Possibly it won't be a problem to share the client instance between (de)serializers, by I opted to avoid it.